feat: cylindrical detector

asalzburger opened the pull request:
This PR adds the infrastructure to create cylindrical detectors with the new `Detector` structure. It adds the helper for cylindrical detector that can do the connection/glueing of portals in R, z, phi and an associated wrapping (with bevelled bounds). These will be needed to build the first `OpenDataDetector` implementation. 

It does also add unit tests for these cases.

The core of this PR is a `CylindricalDetectorHelper`, which is designed to act exclusively on `CylinderVolume` objects and with some extend `BevelledCylinderVolume` objects. It provides the helper methods to e.g. fuse two (or more) cylinder volumes together, which means:
 - share the portals (and transfer portal link information)
 - create new shared/extended portals 
 - bundle this is a proto container information that allows to process this further.

For cylindrical volumes, there are the following options:
 - attach volumes in `binR`, e.g. for concentric barrel volumes
 - attach volumes in `binZ`, e.g. for sequential endocarp volumes
 - attach volumes in `binPhi`, e.g. for phi-sector volumes to be combined

In addition, wrapping cylindrical volumes and proto containers is also supported (using the `BevelledCylinderVolumes` class).

There no or little  focus on checking the consistency, compatibility of the individual volumes in this schema, it is assumed that the incoming objects are already properly synchronised, e.g. through the use of the `ProtoDetector::harmonize()` infrastructure.


@andiwand this is for your information @noemina has agreed to review it.

andiwand commented:
I will also take a look next week :+1: 

maybe we can also discuss how we transition best from the old to the new geometry because to me it looks like this will clash in the algorithms soon

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    // New joint center
+    Acts::Vector3 pCenter = volumeCenter + centerR * refRotation.col(2);
+    transform.pretranslate(pCenter);
+    Acts::ActsScalar halfX =
+        0.5 * (boundValues[Acts::RectangleBounds::BoundValues::eMaxX] -
+               boundValues[Acts::RectangleBounds::BoundValues::eMinX]);
```
could we use `Acts::range_medium(boundaries)` here?

asalzburger replied:
Not on the bound values, because they are not series of the same type.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  Acts::Transform3 transform = Acts::Transform3::Identity();
+  if (binning == Acts::binR) {
+    // Center r calculation
+    Acts::ActsScalar centerR =
+        0.5 * (boundaries[0u] + boundaries[boundaries.size() - 1u]);
```
could we use `Acts::range_medium(boundaries)` here?

asalzburger replied:
here, yes.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  // Autodetector stitch value
+  Acts::BinningValue stitchValue =
+      phiBoundaries.size() == 2u ? Acts::binZ : Acts::binPhi;
+  auto [lengthZ, medZ] = Acts::range_medium(phiBoundaries);
```
```suggestion
  auto [lengthZ, medZ] = Acts::range_medium(zBoundaries);
```

asalzburger replied:
Good catch.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    Acts::Vector3 pCenter = volumeCenter + centerR * refRotation.col(2);
+    transform.pretranslate(pCenter);
+    // New joint bounds
+    Acts::ActsScalar halfY =
+        0.5 * (boundValues[Acts::RectangleBounds::BoundValues::eMaxY] -
+               boundValues[Acts::RectangleBounds::BoundValues::eMinY]);
+    bounds = std::make_unique<Acts::RectangleBounds>(
+        0.5 * (boundaries[boundaries.size() - 1u] - boundaries[0u]), halfY);
```
could we use `Acts::range_medium(boundaries)` here?

asalzburger replied:
not on the boundValues, yes on the boundaries - see above.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    for (const auto& s : selectedSides) {
+      if (pc.find(s) != pc.end()) {
+        auto p = pc.find(s)->second;
+        if (sideVolumes.find(s) == sideVolumes.end()) {
+          sideVolumes[s] = {};
+        }
+        auto& sVolumes = sideVolumes[s];
```
```suggestion
        auto& sVolumes = sideVolumes[s];
```
I think this is the same because the map will construct an empty object if the key does not exist yet

asalzburger replied:
Ok. Updated.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+          sideVolumes[s] = {};
+        }
+        auto& sVolumes = sideVolumes[s];
+        auto aVolumes = Acts::Experimental::attachedDetectorVolumes(*p.get());
```
```suggestion
        auto aVolumes = Acts::Experimental::attachedDetectorVolumes(*p);
```

asalzburger replied:
Ok.

andiwand commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    [[maybe_unused]] const Acts::GeometryContext& gctx,
+    const std::vector<std::shared_ptr<Acts::Experimental::DetectorVolume>>&
+        volumes,
+    const std::vector<std::array<unsigned int, 2u>>& refCur) {
```
```suggestion
    const std::vector<std::pair<unsigned int>>& refCur) {
```

asalzburger replied:
I will leave the array, I like it, don't see what the pair gains me.

andiwand reviewed (commented):
round 2/2

noemina commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    ActsScalar HlZ =
+        volumes[1u]->volumeBounds().values()
+            [Acts::CutoutCylinderVolumeBounds::BoundValues::eHalfLengthZ];
+    ActsScalar innerR =
+        volumes[0u]
+            ->volumeBounds()
+            .values()[Acts::CylinderVolumeBounds::BoundValues::eMinR];
```
Do we assume innerR for the inner volume is the same as for the outer volume? If not, we need to build 3 portals, not only one, don't we?

asalzburger replied:
Yes, we assume everything is synchronised at this step, that's the job of the `ProtoDetector`.

noemina commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    auto sideVolumes = stripSideVolumes({innerContainer}, {3u}, {3u}, logLevel);
+
+    // First the left volume sector
+    std::vector<std::shared_ptr<DetectorVolume>> innerVolumnes = {
```
```suggestion
    std::vector<std::shared_ptr<DetectorVolume>> innerVolumes = {
```

noemina commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    }
+    // Last the right volume sector
+    zBoundaries.push_back(zBoundaries.back() + 2 * pHalfLengthZ);
+    innerVolumnes.push_back(wrappingVolume.getSharedPtr());
```
```suggestion
    innerVolumes.push_back(wrappingVolume.getSharedPtr());
```

asalzburger replied:
Fixed, see above.

noemina reviewed (commented):
I have added suggestions for additional comments and a few questions where the code is not clear.

noemina reviewed (commented):
I have added suggestions for additional comments and a few questions where the code is not clear.

asalzburger commented on Core/include/Acts/Utilities/Helpers.hpp:
```diff
@@ -598,4 +599,46 @@ T clampValue(U value) {
+std::array<typename T::value_type, 2u> min_max(const T& tseries,
+                                               bool sort = true) {
+  if (sort) {
+    T tcopy = tseries;
+    std::sort(tcopy.begin(), tcopy.end());
+    return min_max(tcopy, false);
```
Accepted.

asalzburger commented on Core/include/Acts/Utilities/Helpers.hpp:
```diff
@@ -598,4 +599,46 @@ T clampValue(U value) {
+///
+/// @return [ min, max ] in an array
+template <typename T>
+std::array<typename T::value_type, 2u> min_max(const T& tseries,
```
Will keep with an array, I like array better when it makes clear that they are of the same type.

asalzburger commented on Core/src/Detector/DetectorHelper.cpp:
```diff
@@ -0,0 +1,58 @@
+
+#include "Acts/Detector/CylindricalDetectorHelper.hpp"
+
+#include <exception>
```
Fixed.

asalzburger commented on Tests/UnitTests/Core/Utilities/HelpersTests.cpp:
```diff
@@ -8,10 +8,13 @@
 #include "Acts/Utilities/Helpers.hpp"
 
 #include <bitset>
+#include <climits>
```
Fixed.

asalzburger commented on Core/include/Acts/Detector/DetectorHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+#include "Acts/Utilities/BinningType.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <limits>
+#include <map>
+#include <memory>
+#include <tuple>
+#include <vector>
```
Updated.

asalzburger commented on Core/include/Acts/Detector/CylindricalDetectorHelper.hpp:
```diff
@@ -0,0 +1,158 @@
+#include "Acts/Utilities/Logger.hpp"
+
+#include <limits>
+#include <memory>
```
Fixed.

asalzburger commented on Core/include/Acts/Detector/DetectorHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+class DetectorVolume;
+class Portal;
+
+using ProtoContainer = std::map<unsigned int, std::shared_ptr<Portal>>;
```
Yes, because the CylindricalDetectorHelper is only included in the source file.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+#include "Acts/Utilities/detail/Axis.hpp"
+#include "Acts/Utilities/detail/Grid.hpp"
+
+#include <exception>
```
Fixed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+#include "Acts/Utilities/detail/Grid.hpp"
+
+#include <exception>
+#include <tuple>
+#include <unordered_set>
```
Correct, removed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+               boundValues[Acts::RectangleBounds::BoundValues::eMinX]);
+    // New joint bounds
+    bounds = std::make_unique<Acts::RectangleBounds>(
+        halfX, 0.5 * (boundaries[boundaries.size() - 1u] - boundaries[0u]));
```
Done.

paulgessinger commented:
Could you make the PR title a bit more descriptive?

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Innermost volume boundaries
+  std::vector<ActsScalar> rBoundaries = {};
+  auto refValues = volumes[0u]->volumeBounds().values();
+
+  // Reference boundary values
+  rBoundaries.push_back(refValues[0u]);
+  rBoundaries.push_back(refValues[1u]);
```
I spelled the names out, there's no need for checking, we pick value 0 == eMinR and 1 == eMaxR as a first set of R boundaries.

I changed this to enums though.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  // Check for bounds compatibility
+  // We check outer (1u) vs inner (0u), phi sector (3u) and average phi (4u)
+  std::vector<std::array<unsigned int, 2u>> checks = {
+      {1u, 0u}, {3u, 3u}, {4u, 4u}};
```
I have updated the comment:
```
  // Check for bounds values of volumes (i) and (i+1) succesively for 
  // compatibility:
  // - check outer (1u) of volume (i) vs inner radius (0u) of volume (i+1)
  // - phi sector (3u) and average phi (4u) between volumes (i), (i+1)
```

asalzburger commented on Tests/UnitTests/Core/Detector/CylindricalDetectorHelperTests.cpp:
```diff
@@ -0,0 +1,609 @@
+#include "Acts/Utilities/Enumerate.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <exception>
```
Changed to stdexcept, as we check throw behaviour.

asalzburger commented on Core/include/Acts/Detector/CylindricalDetectorHelper.hpp:
```diff
@@ -0,0 +1,158 @@
+    const std::vector<unsigned int>& selectedOnly = {},
+    Acts::Logging::Level logLevel = Acts::Logging::INFO);
+
+/// @brief Connect detector volumes in phi
```
Fixed.

asalzburger commented on Core/include/Acts/Utilities/Helpers.hpp:
```diff
@@ -598,4 +599,46 @@ T clampValue(U value) {
+template <typename T>
+std::tuple<typename T::value_type, ActsScalar> range_medium(const T& tseries,
+                                                            bool sort = true) {
+  if (sort) {
+    T tcopy(tseries);
+    std::sort(tcopy.begin(), tcopy.end());
+    return range_medium(tcopy, false);
+  }
```
Fixed already, see above.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+}
+
+/// @brief Helper function to create a disc portal replacement
+/// @param gcxt the geometry context of thix call
```
Fixed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    Acts::Logging::Level logLevel = Acts::Logging::INFO) {
+  ACTS_LOCAL_LOGGER(Acts::getDefaultLogger("::stripSideVolumes", logLevel));
+
+  // Thes are the aripped out side volumes
```
Fixed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+        throw std::invalid_argument(message.c_str());
+      }
+      // Set recursively
+      refRotation = curRotation;
```
Right you are, removed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  // We check outer (1u) vs inner (0u), phi sector (3u) and average phi (4u)
+  std::vector<std::array<unsigned int, 2u>> checks = {
+      {1u, 0u}, {3u, 3u}, {4u, 4u}};
+  // And we check the half length if its not a selective attachment
```
Fixed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // If volume sectors are present, these have to be respected
+  if (sectorsPresent) {
+    // Sector assignmenta are forward backward
```
Fixed.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  auto refValues = volumes[0u]->volumeBounds().values();
+
+  // Reference boundary values
+  rBoundaries.push_back(refValues[0u]);
+  rBoundaries.push_back(refValues[1u]);
```
Done already, see above.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  // Fuse the cylinders - portals can be reused for this operation
+  for (unsigned int iv = 1; iv < volumes.size(); ++iv) {
+    refValues = volumes[iv]->volumeBounds().values();
+    rBoundaries.push_back(refValues[1u]);
```
Done.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Proto container setting
+  if (connectR) {
+    if (volumes[0u]->portals().size() == 4u or
+        volumes[0u]->portals().size() == 6u) {
```
Done.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+#include <tuple>
+#include <unordered_set>
+
+namespace {
```
Added explicitly.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    // Only connect if configured to do so
+    if (connectR) {
+      // Fuse and swap
+      auto& keepCylinder = volumes[iv - 1]->portalPtrs()[2u];
+      auto& wasteCylinder = volumes[iv]->portalPtrs()[3u];
```
Added an explicit description.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  }
+
+  // Check if sectors are present by the number of portals
+  bool sectorsPresent = volumes[volumes.size() - 1u]->portals().size() > 4u;
```
Added an explicit description.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Connect in R ?
+  bool connectR = selectedOnly.empty() or
+                  std::find(selectedOnly.begin(), selectedOnly.end(), 2u) !=
```
Added a description.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+      // Fill the map
+      protoContainer[i] = p;
+      // Potential offset for tube vs/ cylinder
+      int iOffset = (iv->portals().size() == 3u and i > 2u) ? -1 : 0;
```
Added.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Connect in Z ?
+  bool connectZ = selectedOnly.empty() or
+                  std::find(selectedOnly.begin(), selectedOnly.end(), 1u) !=
```
Added.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Check if inner cylinder and sectors are present by the number of portals
+  size_t nPortals = volumes[volumes.size() - 1u]->portals().size();
+  bool innerPresent = (nPortals != 3u and nPortals != 5u);
```
That's practically in words what is written in two lines, I think at this stage it is selfexplaining.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+  ACTS_VERBOSE("New combined center calculated at "
+               << toString(combinedCenter));
+
+  std::vector<ActsScalar> zBoundaries = {};
+  for (const auto& zb3D : zBoundaries3D) {
+    auto proj3D = (zb3D - combinedCenter).dot(rotation.col(2));
+    ActsScalar zBoundary =
+        std::copysign((zb3D - combinedCenter).norm(), proj3D);
+    zBoundaries.push_back(zBoundary);
+  }
```
I added a small sentence.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    ACTS_VERBOSE("- update portals of volume '" << iv->name() << "'.");
+    for (auto& [p, i, dir, boundaries, binning] : pReplacements) {
+      // Potential offset for tube vs/ cylinder
+      int iOffset = (i > 2u and not innerPresent) ? -1 : 0;
```
Added.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Positive disc
+  pReplacements.push_back(createDiscReplacement(
+      refTransform, {refValues[0u], refValues[1u]}, phiBoundaries, 1u,
```
Done.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+
+  // Outside cylinder
+  pReplacements.push_back(createCylinderReplacement(
+      refTransform, refValues[1u], {-refValues[2u], refValues[2u]},
```
Also Done.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+      refTransform, refValues[1u], {-refValues[2u], refValues[2u]},
+      phiBoundaries, 2u, Acts::NavigationDirection::Backward));
+
+  if (refValues[0u] > 0.) {
```
Done + comment added.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    std::shared_ptr<Portal> keepCylinder = containers[ic - 1].find(2u)->second;
+    std::shared_ptr<Portal> wasteCylinder = containers[ic].find(3u)->second;
+    keepCylinder->fuse(wasteCylinder);
+    for (auto& av : wasteCylinder->attachedDetectorVolumes()[1u]) {
```
I think it's fine, it is pretty explicit in this way, isn't it?

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    std::shared_ptr<Portal> keepDisc = formerContainer.find(1u)->second;
+    std::shared_ptr<Portal> wasteDisc = currentContainer.find(0u)->second;
+    keepDisc->fuse(wasteDisc);
+    for (auto& av : wasteDisc->attachedDetectorVolumes()[1u]) {
```
Same as above.

asalzburger reviewed (commented):
Answered in comments.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+    Acts::ActsScalar centerR =
+        0.5 * (boundaries[0u] + boundaries[boundaries.size() - 1u]);
+    // New joint center
+    Acts::Vector3 pCenter = volumeCenter + centerR * refRotation.col(2);
```
I hope this makes it clearer.

noemina reviewed (commented):
All comments from my side are resolved.

asalzburger commented:
> 3 comments of mine are hiding in the "show more" thingy. I linked them here. otherwise good to go 🚀
> 
> * [feat: cylindrical detector #1983 (comment)](https://github.com/acts-project/acts/pull/1983#discussion_r1149238668)
> * [feat: cylindrical detector #1983 (comment)](https://github.com/acts-project/acts/pull/1983#discussion_r1149246609)
> * [feat: cylindrical detector #1983 (comment)](https://github.com/acts-project/acts/pull/1983#discussion_r1149248547)

Alright, we should get the changes to the `DetectorVolume`, `DetectorFinder` in first - then adapt this one.

asalzburger commented on Core/src/Detector/CylindricalDetectorHelper.cpp:
```diff
@@ -0,0 +1,1115 @@
+#include "Acts/Utilities/detail/Grid.hpp"
+
+#include <exception>
+#include <tuple>
```
Fixed.

asalzburger commented:
The codecov will go up once the rest of the infrastructure is in and proper testing can be done.

asalzburger commented:
I think I have addressed all comments, please be aware that some follow-up pr #2065 will change location and details a little bit in order to harmonise the new detector building infrastructure. #2065 will be rebased onto this one, once we have it merged in.

In order to avoid merge conflicts, probably we can move this one in (as it is still in the `Experimental` namespace) and pick up eventual remaining points in the follow-up PR which comes also with the appropriate unit testing.

asalzburger commented:
`codecov` will increase with the follow-up PR which allows for coherent testing.
